src/cocotb/share/include/py_gpi_logging.h

```diff
@@ -26,6 +28,47 @@ PYGPILOG_EXPORT void py_gpi_logger_finalize();
 
 extern PYGPILOG_EXPORT PyObject *pEventFn;  // This is gross but I don't care
 
+extern "C" PYGPILOG_EXPORT int pygpi_tracing_enabled;
+extern "C" PYGPILOG_EXPORT int python_context_tracing_enabled;
+extern "C" PYGPILOG_EXPORT int is_python_context;
+
+/** Logs a message at TRACE log level if PYGPI tracing is enabled */
+#define PYGPI_TRACE_MSG(...)         \
+    do {                             \
+        if (pygpi_tracing_enabled) { \
+            LOG_TRACE(__VA_ARGS__);  \
+        }                            \
+    } while (0)
+
+// c_to_python and python_to_c are implemented as macros instead of functions so
+// that the logs reference the user's lineno and filename
+
+#define c_to_python()                                                  \
+    do {                                                               \
+        if (python_context_tracing_enabled) {                          \
+            if (is_python_context) {                                   \
+                LOG_ERROR(                                             \
+                    "FATAL: Trying C => Python but already in Python " \
+                    "context");                                        \
+                exit(1);                                               \
+            }                                                          \
+            ++is_python_context;                                       \
+            LOG_TRACE("C => Python");                                  \
+        }                                                              \
+    } while (0)
+
+#define python_to_c()                                                      \
+    do {                                                                   \
+        if (python_context_tracing_enabled) {                              \
+            if (!is_python_context) {                                      \
+                LOG_ERROR(                                                 \
+                    "FATAL: Trying Python => C but already in C context"); \
+                exit(1);                                                   \
+            }                                                              \
+            --is_python_context;                                           \
+            LOG_TRACE("Python => C");                                      \
+        }                                                                  \
+    } while (0)
```

Comment:
Lets put this in a source file and let LTO do its job.

Reply:
It's used in both the simulator module and libcocotb. I can move it to its own file and include in both libraries and not export it, or we can deal with it later when/if those are libraries are merged.
I don't want to force it to cross the library boundary by only building it into one lib, I would want to encourage it to be inlined by the compiler.